pkg/apis/serving/v1beta1/explainer_custom.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2020 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"github.com/kubeflow/kfserving/pkg/constants"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CustomExplainer defines arguments for configuring a custom transformer.
+type CustomExplainer struct {
+	// This spec is dual purpose.
+	// 1) Users may choose to provide a full PodSpec for their predictor.
+	// The field PodSpec.Containers is mutually exclusive with other Predictors (i.e. TFServing).
+	// 2) Users may choose to provide a Predictor (i.e. TFServing) and specify PodSpec
+	// overrides in the CustomPredictor PodSpec. They must not provide PodSpec.Containers in this case.
+	v1.PodTemplateSpec `json:",inline"`
+}
+
+// Validate returns an error if invalid
+func (c *CustomExplainer) Validate() error {
+	return nil
+}
+
+// Default sets defaults on the resource
+func (c *CustomExplainer) Default(config *InferenceServicesConfig) {
+	c.Name = constants.InferenceServiceContainerName
+	setResourceRequirementDefaults(&c.Spec.Containers[0].Resources)
+}
+
+func (c *CustomExplainer) GetStorageUri() *string {
```

Comment:
Can we genericize this implementation to be the same for all explainers, transformers, predictors. I think having a hierarchy of checking spec.StorageUri and then falling back on a env is a good way of doing this for all explainers.

Reply:
The problem is that `StorageUri` is not at the explainer or predictor spec level, it is a field on specific predictor or explainer spec, maybe we should really move the field one level up and totally get rid of the magic environment variable.